sdks/python/apache_beam/runners/common.py

```diff
@@ -237,6 +239,16 @@ def __init__(self, do_fn):
         method = timer_spec._attached_callback
         self.timer_methods[timer_spec] = MethodWrapper(do_fn, method.__name__)
 
+    self._bundle_finalizer = self.get_bundle_finalizer()
+    if self._bundle_finalizer:
+      self.finalize_method = MethodWrapper(
+          self._bundle_finalizer, 'finalize_bundle')
+
+  def get_bundle_finalizer(self):
```

Comment:
This isn't really a method that we want to call; this seems to be leftover from the previous API. I think we can simplify things as follows: 

1) The DoFnRunner holds an instance of a _BundleFinalizerParam, which it passes to the DoFnInvoker.
2) The PerWindowInvoker swaps this out for core.DoFn.BundleFinalizerParam, the same place it's swapping out ElementParam, WindowParam, TimestampParam, etc.
3) Every Operation has new needs_finalization and finalize_bundle methods, which is a no-op except for ParDoOperation which calls a similar operation on DoFnRunner which looks at whether _BundleFinalizerParam has any callbacks/calls them respectively. 
4) The BundleProcessor object has similar methods, which just walk over its set of all operations and calls them.

Reply:
Changed in the latest version. Thanks!